sw/device/tests/dif/dif_gpio_test.cc

```diff
@@ -0,0 +1,420 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/testing/mock_mmio.h"
+
+extern "C" {
+#include "gpio_regs.h"  // Generated
+#include "sw/device/lib/dif/dif_gpio.h"
+}  // extern "C"
+
+// Base class for the test fixtures in this file.
+class DifGpioTest : public testing::Test, public mock_mmio::MmioTest {};
+
+// Init tests
+class InitTest : public DifGpioTest {};
+
+TEST_F(InitTest, NullArgs) {
+  dif_gpio_config_t config{.base_addr = dev().region()};
+  dif_gpio_t out_arg_gpio;
+
+  EXPECT_EQ(dif_gpio_init(nullptr, &out_arg_gpio),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_init(&config, nullptr), kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_init(nullptr, nullptr), kDifGpioResultInvalidArgument);
+}
+
+TEST_F(InitTest, Init) {
+  EXPECT_WRITE32(GPIO_INTR_ENABLE_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_DIRECT_OE_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_DIRECT_OUT_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_INTR_CTRL_EN_RISING_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_INTR_CTRL_EN_FALLING_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_INTR_CTRL_EN_LVLHIGH_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_INTR_CTRL_EN_LVLLOW_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_CTRL_EN_INPUT_FILTER_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_INTR_STATE_REG_OFFSET, 0xFFFFFFFFu);
+
+  dif_gpio_config_t config{.base_addr = dev().region()};
+  dif_gpio_t gpio;
+  EXPECT_EQ(dif_gpio_init(&config, &gpio), kDifGpioResultOK);
+}
+
+// Base class for the rest of the tests in this file, provides a
+// `dif_gpio_t` instance.
+class DifGpioTestWithDevice : public DifGpioTest {
+ protected:
+  uint32_t AllZerosExcept(uint32_t index) { return 1 << index; }
+
+  uint32_t AllOnesExcept(uint32_t index) { return ~AllZerosExcept(index); }
+
+  dif_gpio_t const gpio_ = {.base_addr = dev().region()};
```

Comment:
I would actually suggest avoiding `const` members in C++ altogether: they cause all sorts of issues with move constructors.

Reply:
Done (old habits die hard, I guess).

@mcy, yes they do, but IMHO using `const` in this case is reasonable because having an API that enforces this would be an overkill given the close relationship between test fixtures and tests.